File: WaterTracker.java
// WaterTracker.java
import java.io.*;
import java.nio.file.*;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.util.stream.Collectors;

record WaterEntry(int id, String timestamp, int amount, String note) implements Serializable {}

class WaterData implements Serializable {
    private static final long serialVersionUID = 1L;
    public int dailyGoal;
    public List<WaterEntry> entries;
}

class WaterTracker implements Serializable {
    private static final long serialVersionUID = 1L;
    private List<WaterEntry> entries = new ArrayList<>();
    private int nextId = 1;
    private int dailyGoal = 2000;

    public WaterEntry addEntry(int amount, String note) {
        if (amount <= 0) throw new IllegalArgumentException("Объём должен быть положительным");
        WaterEntry entry = new WaterEntry(
            nextId,
            LocalDateTime.now().format(DateTimeFormatter.ISO_LOCAL_DATE_TIME),
            amount,
            note != null ? note : ""
        );
        entries.add(entry);
        nextId++;
        return entry;
    }

    public Optional<WaterEntry> findEntry(int id) {
        return entries.stream().filter(e -> e.id() == id).findFirst();
    }

    public boolean deleteEntry(int id) {
        return entries.removeIf(e -> e.id() == id);
    }

    public List<WaterEntry> getTodayEntries() {
        LocalDateTime today = LocalDateTime.now().withHour(0).withMinute(0).withSecond(0).withNano(0);
        return entries.stream()
            .filter(e -> LocalDateTime.parse(e.timestamp()).isAfter(today) || LocalDateTime.parse(e.timestamp()).isEqual(today))
            .collect(Collectors.toList());
    }

    public List<WaterEntry> getWeekEntries() {
        LocalDateTime weekAgo = LocalDateTime.now().minusDays(7);
        return entries.stream()
            .filter(e -> LocalDateTime.parse(e.timestamp()).isAfter(weekAgo) || LocalDateTime.parse(e.timestamp()).isEqual(weekAgo))
            .collect(Collectors.toList());
    }

    public Map<String, Object> getStats() {
        List<WaterEntry> today = getTodayEntries();
        List<WaterEntry> week = getWeekEntries();
        int todayTotal = today.stream().mapToInt(WaterEntry::amount).sum();
        int weekTotal = week.stream().mapToInt(WaterEntry::amount).sum();
        double avgDaily = weekTotal / 7.0;
        double progress = dailyGoal > 0 ? (double) todayTotal / dailyGoal * 100 : 0;
        if (progress > 100) progress = 100;
        Map<String, Object> stats = new HashMap<>();
        stats.put("today_total", todayTotal);
        stats.put("today_count", today.size());
        stats.put("week_total", weekTotal);
        stats.put("week_count", week.size());
        stats.put("avg_daily", avgDaily);
        stats.put("daily_goal", dailyGoal);
        stats.put("progress", progress);
        return stats;
    }

    public void setDailyGoal(int goal) {
        if (goal <= 0) throw new IllegalArgumentException("Цель должна быть положительной");
        dailyGoal = goal;
    }

    public void saveToFile(String filename) throws IOException {
        WaterData data = new WaterData();
        data.dailyGoal = dailyGoal;
        data.entries = new ArrayList<>(entries);
        try (ObjectOutputStream oos = new ObjectOutputStream(Files.newOutputStream(Paths.get(filename)))) {
            oos.writeObject(data);
        }
    }

    public void loadFromFile(String filename) throws IOException, ClassNotFoundException {
        try (ObjectInputStream ois = new ObjectInputStream(Files.newInputStream(Paths.get(filename)))) {
            WaterData data = (WaterData) ois.readObject();
            dailyGoal = data.dailyGoal;
            entries = new ArrayList<>(data.entries);
            for (WaterEntry e : entries) {
                if (e.id() >= nextId) nextId = e.id() + 1;
            }
        }
    }

    public List<WaterEntry> getEntries() { return Collections.unmodifiableList(entries); }
}

public class WaterTrackerApp {
    private static final Scanner scanner = new Scanner(System.in);

    private static String readString(String prompt) {
        System.out.print(prompt);
        return scanner.nextLine().trim();
    }

    private static int readInt(String prompt) {
        while (true) {
            try {
                System.out.print(prompt);
                return Integer.parseInt(scanner.nextLine().trim());
            } catch (NumberFormatException e) {
                System.out.println("Введите число.");
            }
        }
    }

    private static void printEntry(WaterEntry entry) {
        System.out.printf("#%d - %s - %d мл%n", entry.id(), entry.timestamp(), entry.amount());
        if (!entry.note().isEmpty()) {
            System.out.printf("   Заметка: %s%n", entry.note());
        }
    }

    public static void main(String[] args) {
        WaterTracker tracker = new WaterTracker();
        try {
            tracker.loadFromFile("water_data.ser");
        } catch (IOException | ClassNotFoundException e) {
            System.out.println("Не удалось загрузить данные, начинаем с пустого трекера.");
        }

        while (true) {
            System.out.println("\n===== ТРЕКЕР ВОДНОГО БАЛАНСА (Java) =====");
            System.out.println("1. Добавить запись");
            System.out.println("2. Просмотреть записи за сегодня");
            System.out.println("3. Просмотреть записи за неделю");
            System.out.println("4. Показать статистику");
            System.out.println("5. Установить дневную цель");
            System.out.println("6. Удалить запись");
            System.out.println("7. Сохранить в файл");
            System.out.println("8. Загрузить из файла");
            System.out.println("0. Выход");
            String choice = readString("Выберите действие: ");

            switch (choice) {
                case "0" -> { return; }
                case "1" -> {
                    int amount = readInt("Объём (мл): ");
                    if (amount <= 0) {
                        System.out.println("Объём должен быть положительным.");
                        continue;
                    }
                    String note = readString("Заметка (необязательно): ");
                    WaterEntry entry = tracker.addEntry(amount, note);
                    System.out.printf("Запись добавлена с ID %d%n", entry.id());
                }
                case "2" -> {
                    var entries = tracker.getTodayEntries();
                    if (entries.isEmpty()) {
                        System.out.println("За сегодня записей нет.");
                    } else {
                        System.out.printf("\nЗаписи за сегодня (%d):%n", entries.size());
                        entries.forEach(WaterTrackerApp::printEntry);
                    }
                }
                case "3" -> {
                    var entries = tracker.getWeekEntries();
                    if (entries.isEmpty()) {
                        System.out.println("За неделю записей нет.");
                    } else {
                        System.out.printf("\nЗаписи за последние 7 дней (%d):%n", entries.size());
                        entries.forEach(WaterTrackerApp::printEntry);
                    }
                }
                case "4" -> {
                    var stats = tracker.getStats();
                    System.out.println("\n=== СТАТИСТИКА ===");
                    System.out.printf("Дневная цель: %d мл%n", stats.get("daily_goal"));
                    System.out.printf("Сегодня выпито: %d мл (%d записей)%n", stats.get("today_total"), stats.get("today_count"));
                    System.out.printf("Прогресс: %.1f%%%n", stats.get("progress"));
                    System.out.printf("За неделю: %d мл (%d записей)%n", stats.get("week_total"), stats.get("week_count"));
                    System.out.printf("Среднее за день (за неделю): %.1f мл%n", stats.get("avg_daily"));
                }
                case "5" -> {
                    int goal = readInt("Новая дневная цель (мл): ");
                    try {
                        tracker.setDailyGoal(goal);
                        System.out.printf("Цель установлена: %d мл%n", goal);
                    } catch (IllegalArgumentException e) {
                        System.out.println(e.getMessage());
                    }
                }
                case "6" -> {
                    int id = readInt("Введите ID записи для удаления: ");
                    if (tracker.deleteEntry(id)) {
                        System.out.println("Запись удалена.");
                    } else {
                        System.out.println("Запись не найдена.");
                    }
                }
                case "7" -> {
                    try {
                        tracker.saveToFile("water_data.ser");
                        System.out.println("Сохранено.");
                    } catch (IOException e) {
                        System.out.println("Ошибка сохранения: " + e.getMessage());
                    }
                }
                case "8" -> {
                    try {
                        tracker.loadFromFile("water_data.ser");
                        System.out.println("Загружено.");
                    } catch (IOException | ClassNotFoundException e) {
                        System.out.println("Ошибка загрузки: " + e.getMessage());
                    }
                }
                default -> System.out.println("Неизвестная команда.");
            }
        }
    }
}

File: WaterTracker.cs
// WaterTracker.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;

public record WaterEntry(
    int Id,
    string Timestamp,
    int Amount,
    string Note
);

public class WaterData
{
    public int DailyGoal { get; set; }
    public List<WaterEntry> Entries { get; set; } = new();
}

public class WaterTracker
{
    private List<WaterEntry> entries = new();
    private int nextId = 1;
    private int dailyGoal = 2000;

    public IReadOnlyList<WaterEntry> Entries => entries.AsReadOnly();
    public int DailyGoal => dailyGoal;

    public WaterEntry AddEntry(int amount, string note = "")
    {
        if (amount <= 0) throw new ArgumentException("Объём должен быть положительным");
        var entry = new WaterEntry(
            nextId,
            DateTime.Now.ToString("o"),
            amount,
            note ?? ""
        );
        entries.Add(entry);
        nextId++;
        return entry;
    }

    public WaterEntry? FindEntry(int id) => entries.FirstOrDefault(e => e.Id == id);

    public bool DeleteEntry(int id) => entries.RemoveAll(e => e.Id == id) > 0;

    public List<WaterEntry> GetTodayEntries()
    {
        var today = DateTime.Today;
        return entries.Where(e => DateTime.Parse(e.Timestamp).Date == today).ToList();
    }

    public List<WaterEntry> GetWeekEntries()
    {
        var weekAgo = DateTime.Now.AddDays(-7);
        return entries.Where(e => DateTime.Parse(e.Timestamp) >= weekAgo).ToList();
    }

    public Dictionary<string, object> GetStats()
    {
        var today = GetTodayEntries();
        var week = GetWeekEntries();
        int todayTotal = today.Sum(e => e.Amount);
        int weekTotal = week.Sum(e => e.Amount);
        double avgDaily = weekTotal / 7.0;
        double progress = dailyGoal > 0 ? (double)todayTotal / dailyGoal * 100 : 0;
        if (progress > 100) progress = 100;
        return new Dictionary<string, object>
        {
            ["today_total"] = todayTotal,
            ["today_count"] = today.Count,
            ["week_total"] = weekTotal,
            ["week_count"] = week.Count,
            ["avg_daily"] = avgDaily,
            ["daily_goal"] = dailyGoal,
            ["progress"] = progress
        };
    }

    public void SetDailyGoal(int goal)
    {
        if (goal <= 0) throw new ArgumentException("Цель должна быть положительной");
        dailyGoal = goal;
    }

    public void SaveToFile(string filename)
    {
        var data = new WaterData { DailyGoal = dailyGoal, Entries = entries };
        var options = new JsonSerializerOptions { WriteIndented = true };
        string json = JsonSerializer.Serialize(data, options);
        File.WriteAllText(filename, json);
    }

    public void LoadFromFile(string filename)
    {
        if (!File.Exists(filename)) return;
        string json = File.ReadAllText(filename);
        var data = JsonSerializer.Deserialize<WaterData>(json);
        if (data != null)
        {
            dailyGoal = data.DailyGoal;
            entries = data.Entries;
            nextId = entries.Any() ? entries.Max(e => e.Id) + 1 : 1;
        }
    }
}

public static class Program
{
    private static string ReadString(string prompt)
    {
        Console.Write(prompt);
        return Console.ReadLine()?.Trim() ?? "";
    }

    private static int ReadInt(string prompt)
    {
        while (true)
        {
            Console.Write(prompt);
            if (int.TryParse(Console.ReadLine(), out int result))
                return result;
            Console.WriteLine("Введите число.");
        }
    }

    private static void PrintEntry(WaterEntry entry)
    {
        Console.WriteLine($"#{entry.Id} - {entry.Timestamp} - {entry.Amount} мл");
        if (!string.IsNullOrEmpty(entry.Note))
            Console.WriteLine($"   Заметка: {entry.Note}");
    }

    public static void Main()
    {
        var tracker = new WaterTracker();
        try { tracker.LoadFromFile("water_data.json"); }
        catch { Console.WriteLine("Не удалось загрузить данные."); }

        while (true)
        {
            Console.WriteLine("\n===== ТРЕКЕР ВОДНОГО БАЛАНСА (C#) =====");
            Console.WriteLine("1. Добавить запись");
            Console.WriteLine("2. Просмотреть записи за сегодня");
            Console.WriteLine("3. Просмотреть записи за неделю");
            Console.WriteLine("4. Показать статистику");
            Console.WriteLine("5. Установить дневную цель");
            Console.WriteLine("6. Удалить запись");
            Console.WriteLine("7. Сохранить в файл");
            Console.WriteLine("8. Загрузить из файла");
            Console.WriteLine("0. Выход");
            string choice = ReadString("Выберите действие: ");

            switch (choice)
            {
                case "0": return;
                case "1":
                    int amount = ReadInt("Объём (мл): ");
                    if (amount <= 0) { Console.WriteLine("Объём должен быть положительным."); continue; }
                    string note = ReadString("Заметка (необязательно): ");
                    var entry = tracker.AddEntry(amount, note);
                    Console.WriteLine($"Запись добавлена с ID {entry.Id}");
                    break;
                case "2":
                    var today = tracker.GetTodayEntries();
                    if (today.Count == 0) Console.WriteLine("За сегодня записей нет.");
                    else { Console.WriteLine($"\nЗаписи за сегодня ({today.Count}):"); today.ForEach(PrintEntry); }
                    break;
                case "3":
                    var week = tracker.GetWeekEntries();
                    if (week.Count == 0) Console.WriteLine("За неделю записей нет.");
                    else { Console.WriteLine($"\nЗаписи за последние 7 дней ({week.Count}):"); week.ForEach(PrintEntry); }
                    break;
                case "4":
                    var stats = tracker.GetStats();
                    Console.WriteLine("\n=== СТАТИСТИКА ===");
                    Console.WriteLine($"Дневная цель: {stats["daily_goal"]} мл");
                    Console.WriteLine($"Сегодня выпито: {stats["today_total"]} мл ({stats["today_count"]} записей)");
                    Console.WriteLine($"Прогресс: {stats["progress"]:F1}%");
                    Console.WriteLine($"За неделю: {stats["week_total"]} мл ({stats["week_count"]} записей)");
                    Console.WriteLine($"Среднее за день (за неделю): {stats["avg_daily"]:F1} мл");
                    break;
                case "5":
                    int goal = ReadInt("Новая дневная цель (мл): ");
                    try { tracker.SetDailyGoal(goal); Console.WriteLine($"Цель установлена: {goal} мл"); }
                    catch (Exception ex) { Console.WriteLine(ex.Message); }
                    break;
                case "6":
                    int id = ReadInt("Введите ID записи для удаления: ");
                    if (tracker.DeleteEntry(id)) Console.WriteLine("Запись удалена.");
                    else Console.WriteLine("Запись не найдена.");
                    break;
                case "7":
                    try { tracker.SaveToFile("water_data.json"); Console.WriteLine("Сохранено."); }
                    catch (Exception ex) { Console.WriteLine($"Ошибка: {ex.Message}"); }
                    break;
                case "8":
                    try { tracker.LoadFromFile("water_data.json"); Console.WriteLine("Загружено."); }
                    catch (Exception ex) { Console.WriteLine($"Ошибка: {ex.Message}"); }
                    break;
                default: Console.WriteLine("Неизвестная команда."); break;
            }
        }
    }
}
